### PR TITLE
Add Riemannian optimizers for mixed U(2)/U(1)^4 manifolds

### LINKED ISSUE #24: Adding Batch and GPU

### LINKED ISSUE #29: Feat: Riemannian Optimization Alternative
Give other options for the optimizer. Riemannian Gradient Descent might not be so effective after visualization of the training loss. Shown in PR #27  Considering other options for the optimizer.

### LINKED ISSUE #34: Feat: enhance the Manopt.jl and Manifold.jl with GPU feature
Issue is https://github.com/JuliaManifolds/Manopt.jl/issues/572 here but I would like to fix the plan for implementation. I am still a novice in the GPU development for tensor network . Could we implement in the following steps for the optimizer? Using vibe coding:

1. Fork and create a GPU feature branch 
2. Make the Agent know how the extension work, using ext/CUDA.jl  
3. Make the Agent know how the OMEinsum.jl works. Especially the cuda feature in the package. 
4. Make the Agent check the existing tests and write GPU version of tests and comparison. 
5. Make the Agent implements GPU version of the Manopt.jl in the extension. Since it depends on the Manifold.jl, might also need the extension on the Manifold.jl
6. Ensure extended GPU version of Manopt.jl works the same level as the CPU version, with 20x speed up. 

I am not sure whether I missed anything in the plan, could you help me confirm the steps work? After implementation on the optimizer, I will again use it for the batch and GPU part of the code for ParametricDFT. @GiggleLiu @claude

### LINKED ISSUE #39: Fix: code redundancy
#35 The optimization code with batch and GPU is shown working with test, GPU version is 20x faster than the 1 batch CPU version for the training. Now, the code need to be reduced to the level where review is available, both for the code content and the tests.

### PR DESCRIPTION
**Pull request by zazabap**
## Summary

Refactor the monolithic Riemannian optimizer into a modular, extensible architecture with custom GPU-friendly implementations, replacing external Manifolds.jl/Manopt.jl dependencies for optimization.

### Architecture

| Before | After |
|--------|-------|
| Single 599-line `riemannian_optimizers.jl` | `manifolds.jl` (242 lines) + `optimizers.jl` (393 lines) |
| Hardcoded 2x2 matrix operations | Generic `AbstractRiemannianManifold` with batched `(d1,d2,n)` arrays |
| Manifolds.jl/Manopt.jl as runtime deps | Custom implementations; Manifolds/Manopt test-only (cross-validation) |
| Separate `batched_einsum.jl` | Merged into `loss.jl` |
| Dual Manopt/custom-GPU training paths | Single `optimize\!` code path |

### Key Components

**`src/manifolds.jl`** — Device-agnostic manifold abstraction
- `UnitaryManifold` (U(n)): skew-Hermitian projection, modified Gram-Schmidt retraction
- `PhaseManifold` (U(1)^d): phase projection, normalize retraction
- Batched operations via broadcasting (works on CPU and GPU without cuSOLVER)

**`src/optimizers.jl`** — Two production optimizers
- `RiemannianGD`: gradient descent with Armijo backtracking line search
- `RiemannianAdam`: Becigneul & Ganea (2019) with momentum transport
- NVTX profiling hooks for GPU instrumentation

**`src/training.jl`** — Simplified pipeline
- Single `optimize\!` call replaces branching Manopt/GPU logic
- Batched einsum for parallel multi-image processing
- Device abstraction (`to_device(:cpu/:gpu)`)

### Issues Addressed

Closes #24, closes #29, closes #34, closes #39

Addresses #28, #41, #40

Supersedes #30

## Test Plan

- [x] 980 tests pass (`Pkg.test()`)
- [x] Manifold unit tests: project/retract/transport correctness, unitarity preservation
- [x] Optimizer tests: GD/Adam loss reduction, cross-optimizer consistency, manifold constraint preservation
- [x] Manopt.jl cross-check: our GD/Adam vs Manopt `gradient_descent` on same problem (within 5x)
- [x] Batched einsum: L1/L2/MSE match per-image computation, Zygote gradients match
- [x] All circuit types: QFT, EntangledQFT, TEBD

### New
- `src/manifolds.jl` — Manifold abstraction layer
- `src/optimizers.jl` — RiemannianGD + RiemannianAdam
- `test/loss_tests.jl` — Loss + batched einsum tests (replaces `losstests.jl`)
- `test/manifold_tests.jl` — Manifold unit tests
- `test/optimizer_tests.jl` — Optimizer + Manopt cross-check tests
- `test/cuda_tests.jl` — GPU tests

### Modified
- `src/training.jl` — Single optimize\! code path
- `src/loss.jl` — Merged batched einsum operations
- `src/ParametricDFT.jl` — Updated exports, removed Manifolds/Manopt imports
- `src/qft.jl` — Removed `generate_manifold`, `tensors2point`, `point2tensors`
- `src/basis.jl` — Removed `get_manifold`
- `ext/CUDAExt.jl` — NVTX profiling, device transfer
- `Project.toml` — Manifolds/Manopt moved to `[extras]`

### Deleted
- `test/losstests.jl` — Replaced by `test/loss_tests.jl`

### CODE REVIEW

File: src/batched_einsum.jl [hunk 1]
```diff
@@ -0,0 +1,91 @@
+# Batched einsum: add a batch dimension to einsum contractions so B images
+# are processed in a single kernel call (B×K launches → K launches).
+
```
Comment (zazabap) [inline src/batched_einsum.jl]: The dimension and batch einsum, could they be integrated into the loss.jl for batched version?

**Review by zazabap — COMMENTED** [review]
The redundancy of the code is terrible.

File: src/loss.jl [hunk 2]
```diff
@@ -145,7 +145,7 @@ end
 # ============================================================================
 
 """
```
Comment (zazabap) [inline src/loss.jl]: Docstring here should be simplied. every parameter in the loss function should be crystal clear

File: src/manifolds.jl [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+end
+
+"""
+    classify_manifold(t::AbstractMatrix) -> AbstractRiemannianManifold
+
+Classify a single tensor into its manifold type based on unitarity check.
+Must be called on CPU arrays (use `Array(t)` for GPU tensors).
+"""
+function classify_manifold end  # forward declaration
+
```
Comment (zazabap) [inline src/manifolds.jl]: https://github.com/CodingThrust/problem-reductions/blob/main/.claude/skills/review-implementation/quality-reviewer-prompt.md

File: src/optimizers.jl [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+    if any(g -> any(x -> isnan(x) || isinf(x), g), euclidean_grads)
+        verbose && println("  WARNING: NaN or Inf in gradients at iter $iter")
+        return nothing
+    end
+
+    return euclidean_grads
+end
+
+"""
+    _batched_project(manifold_groups, point_batches, grad_buf_batches, euclidean_grads)
```
Comment (zazabap) [inline src/optimizers.jl]: Will this batch option influence the overall structure of the manifold? How are the batch arranged on the GPUs?

File: src/qft.jl [hunk 5]
```diff
@@ -105,25 +66,28 @@ theta = fft_with_training(m, n, pic, MSELoss(100); steps=200)
...
-    )
-    
-    return result
+
+    # Convert tensors to Matrix{ComplexF64} for optimize!
+    mat_tensors = Matrix{ComplexF64}[Matrix{ComplexF64}(t) for t in tensors]
+
+    loss_fn = ts -> loss_function(ts, m, n, optcode, pic, loss; inverse_code=inverse_code)
+    grad_fn = ts -> begin
+        _, back = Zygote.pullback(loss_fn, ts)
```
Comment (zazabap) [inline src/qft.jl]: Is this Zygote working well?

File: src/training.jl [hunk 6]
```diff
@@ -117,26 +216,28 @@ function _train_basis_core(
             println("  Avg train loss: $(round(avg_train_loss, digits=6))")
             println("  Validation loss: $(round(val_loss, digits=6))")
         end
-        
+
         # Check for improvement
         if val_loss < best_val_loss
             improvement = best_val_loss == Inf ? 100.0 : (best_val_loss - val_loss) / best_val_loss * 100
             verbose && println("  ✓ Validation improved by $(round(improvement, digits=2))%")
             best_val_loss = val_loss
```
Comment (zazabap) [inline src/training.jl]: why use copy. here?

**Review by zazabap — COMMENTED** [review]
Here are some comments for the source code part, which does not write concise code easy to maintain and extend.
